src/handle.rs

```diff
@@ -432,3 +449,356 @@ impl HandleHost {
+  /// cannot be used with this method again. Additionally, it is unsound to call
+  /// this method with an isolate other than that in which the original `Weak`
+  /// was created.
+  pub unsafe fn from_raw(
+    isolate: &mut Isolate,
+    data: Option<NonNull<WeakData<T>>>,
+  ) -> Self {
+    Weak {
+      data: data.map(|raw| Box::from_raw(raw.cast().as_ptr())),
+      isolate_handle: isolate.thread_safe_handle(),
+    }
+  }
+
+  /// Consume this `Weak` handle and return the underlying raw pointer, or
+  /// `None` if the value has been GC'd.
+  ///
+  /// The return value can be converted back into a `Weak` by using
+  /// [`Weak::from_raw`]. Note that `Weak` allocates some memory, and if this
+  /// method returns `Some`, the pointer must be converted back into a `Weak`
+  /// for it to be freed.
+  ///
+  /// Note that this method might return `Some` even after the V8 value has been
+  /// GC'd.
+  pub fn into_raw(mut self) -> Option<NonNull<WeakData<T>>> {
+    if let Some(data) = self.data.take() {
+      let has_finalizer = {
+        let finalizer = data.finalizer.take();
+        let has_finalizer = finalizer.is_some();
+        data.finalizer.set(finalizer);
+        has_finalizer
+      };
+
+      if data.pointer.get().is_none() && !has_finalizer {
+        // If the pointer is None and we're not waiting for the second pass,
+        // drop the box and return None.
+        None
+      } else {
+        assert!(!data.weak_dropped.get());
+        Some(unsafe { NonNull::new_unchecked(Box::into_raw(data)) })
+      }
+    } else {
+      None
+    }
+  }
+
+  fn get_pointer(&self) -> Option<NonNull<T>> {
+    if let Some(data) = &self.data {
+      // It seems like when the isolate is dropped, even the first pass callback
+      // might not be called.
+      if unsafe { self.isolate_handle.get_isolate_ptr() }.is_null() {
+        None
+      } else {
+        data.pointer.get()
+      }
+    } else {
+      None
+    }
+  }
+
+  pub fn is_empty(&self) -> bool {
+    self.get_pointer().is_none()
+  }
+
+  pub fn to_global(&self, isolate: &mut Isolate) -> Option<Global<T>> {
+    if let Some(data) = self.get_pointer() {
+      let handle_host: HandleHost = (&self.isolate_handle).into();
+      handle_host.assert_match_isolate(isolate);
+      Some(unsafe { Global::from_raw(isolate, data) })
+    } else {
+      None
+    }
+  }
+
+  pub fn to_local<'s>(
+    &self,
+    scope: &mut HandleScope<'s, ()>,
+  ) -> Option<Local<'s, T>> {
+    if let Some(data) = self.get_pointer() {
+      let handle_host: HandleHost = (&self.isolate_handle).into();
+      handle_host.assert_match_isolate(scope);
+      let local = unsafe {
+        scope.cast_local(|sd| {
+          v8__Local__New(sd.get_isolate_ptr(), data.cast().as_ptr()) as *const T
+        })
+      };
+      Some(local.unwrap())
+    } else {
+      None
+    }
+  }
+
+  pub fn unset_finalizer(&mut self) {
```

Comment:
This method doesn't really have to be `mut`, since the finalizer is behind a `Cell`, but it semantically does mutate the handle. WDYT?

Reply:
I think this is better that using `&self` 👍 let's keep it